Allow customizing chat message format

ClobberXD opened the pull request:
See https://github.com/minetest/minetest/pull/8021#issuecomment-487998164 and successive comments.

Chat message format (`<name> message`) used to be hardcoded in `server.cpp`. Now the chat message format can be customised by setting `chat_message_format` in `minetest.conf`. This setting takes a string which is the format of the message, and supports the following placeholders:

- `@name` - Name of message sender
- `@message` - Message
- `@timestamp` - (Optional) Timestamp (HH:MM:SS)

`@name` and `@message` are mandatory, for obvious reasons. This setting defaults to `<@name> @message` to retain current behaviour.

### Testing

- Add the following line to your `minetest.conf`:

  ```
  chat_message_format = [@timestamp] @name: @message
  ```

- Chat messages should be displayed like so:

  ```
  [12:34:56] singleplayer: Hello world!
  ```

### Screenshots

- `chat_message_format = [@name]: @message`

  ![screenshot_20190513_201119](https://user-images.githubusercontent.com/36130650/57631585-637e4880-75bd-11e9-8d47-d1320e3338d7.png)

- `chat_message_format = [@timestamp] @name: @message`

  ![screenshot_20190514_164845](https://user-images.githubusercontent.com/36130650/57694060-684c0680-7668-11e9-94e2-5e5570220b4a.png)

### Use-cases

- Customizability
- #8021

****

Tested. Everything works.

ClobberXD commented:
Finished testing. Everything seems to work after some minor tweaks. I've also added a screenshot to the first post.

Please review, approve, and merge. The changes are small and the code is fairly simple. :)

DS-Minetest commented:
Is it possible to have a player name like `"myname%mbla"`?

HybridDog commented:
> […] with player name and message represented by placeholders %n and %m respectively.

I think you could use %name and %message as placeholders; these are more straightforward to understand.

ClobberXD commented:
> Is it possible to have a player name like "myname%mbla"?

Not possible, even in master - player names only support alphanumeric chars, and `-` and `_`.

> And what is if someone wants to have a format like "%m player_name %m msg"?

Splitting name or message across multiple place-holders shouldn't be possible, and doesn't make any sense. How would the message be split between the two placeholders? The code currently recognizes and replaces the first occurrence of the place-holders, and the second one would be left untouched. Good question though :)

> I think you could use %name and %message as placeholders; these are more straightforward to understand.

That's a good idea. Will do.

sorcerykid commented:
Cool idea. But what about support for timestamps? :)

Also can this implementation be extended? Or is it hardcoded in the CPP side?

ClobberXD commented:
This can't be extended to recognize other placeholders from Lua, but I could add support for timestamps in the C++ code itself. That's actually a very useful idea. :)

ClobberXD commented:
Added support for optional timestamp placeholder `%timestamp`, thanks to @sorcerykid for the idea. :)

Updated first post and added another screenshot showing chat messages with timestamps.

rubenwardy commented:
I'd prefer this to be implemented by a new Lua function called `core.format_chat_message(name, msg)`, because it can then be overridden by things like CTC

DS-Minetest commented:
> > And what is if someone wants to have a format like "%m player_name %m msg"?
> 
> Splitting name or message across multiple place-holders shouldn't be possible, and doesn't make any sense. How would the message be split between the two placeholders? The code currently recognizes and replaces the first occurrence of the place-holders, and the others would be left untouched. Good question though :)

What I meant here was something else:
What is if someone wants the chat to look like this:
```
%m ANAND %m: Hi.
%m DS %m: Hello.
%m DS %m: Maybe the %m things in your PR should be escapeable.
```
Instead of like this:
```
<ANAND> Hi.
<DS> Hello.
<DS> Maybe the %m things in your PR should be escapeable.
```

> I'd prefer this to be implemented by a new Lua function called `core.format_chat_message(name, msg)`, because it can then be overridden by things like CTF

:+1: 
Colorizing and different formats for different players are important.

ClobberXD commented:
Ah now I understand what you mean. Regarding %m and %n being left as-is, I've replaced `n` and `m` with full words - `%name` and `%message` as suggested by HybridDog. Now, there's a much lesser chance of conflicts with unintended occurrences of the placeholders. The chance of conflicts with intentionally placed placeholders (e.g. `%name singleplayer | %message Hello`) is extremely remote, and I don't think adding placeholder escapes is worth the effort.

ClobberXD commented:
@rubenwardy How does MT call Lua functions from C++? You asked me to check how `is_protected` is called from C++, but it seems like it isn't called at all.

ClobberXD commented:
Found an article that explains how Lua functions can be called from C++: http://gamedevgeek.com/tutorials/calling-lua-functions/

ClobberXD commented:
Updated PR:

- Chat message formatting is now implemented in Lua. `core.format_chat_message` - takes a name and a message, and returns a formatted string. The most suitable place for defining this function is `/builtin/game/chatcommands.lua` as that implements all chat-related stuff (`/builtin/common/chatcommands.lua` does too, but this seemed better suited). But since this file now contains `core.format_chat_message` which has nothing to do with chat commands, the file has been renamed to `chat.lua`.
- Since `%` is a special character in Lua, the placeholder prefix is now `@` (so `@name`, `@message`, and `@timestamp`), for the sake of simplicity.
- The format of the timestamp has changed from `YYYY-MM-DD HH:MM:SS` to `HH:MM:SS`.

I haven't updated the screenshots, as they're still valid.

I have plans to implement more placeholders, but I'll do it in another PR in the future. This PR is ready for review :)

P.S. @rubenwardy

DS-Minetest commented:
- I don't think that the setting is still needed.
- Imo it should be possible to not send any message by returning `nil` or an empty string or whatever.

ClobberXD commented:
- Good point. However I'm not quite sure about this, so I'll wait for more inputs.

- This function only formats the chat messages, and has nothing to do with allowing/disallowing messages. It's better to retain this distinction. We already have a way to restrict sending chat messages (`register_on_chat_message`), and adding more ways to do the same would just result in confusion.

- Will do, thanks.

DS-Minetest commented:
`minetest.format_chat_message` is still not documented.

(Although the setting is somewhat unneeded, it doesn't hurt, so I don't care.)

ClobberXD commented:
Added documentation to `lua_api.txt`. I'm sure it can be improved, and some suggestions would be great.

paramat commented on src/server.cpp:
```diff
@@ -2863,28 +2863,28 @@ std::wstring Server::handleChat(const std::string &name, const std::wstring &wna
+			std::wstringstream ws;
+			ws << L"You cannot send more messages. You are limited to "
+					<< g_settings->getFloat("chat_message_limit_per_10sec")
+					<< L" messages per 10 seconds.";
```
1 indent is enough

SmallJoker replied:
The code style allows that and LINT would probably have issues with any other notation due to its limited configuration set.

ClobberXD replied:
@paramat Our code style demands double-tab indents for continuation lines, IIRC.

paramat replied:
Sorry, as long as it's readable any amount of indentation is ok.

Fuck what LINT thinks, 95% of it's requests are ridiculous =)
We're not going to make a mess of MT code just because LINT is crap.

> Our code style demands double-tab indents for continuation lines, IIRC.

I hope not, if it does that needs editing, 1 is fine except when the indented line would be aligned with a following indented line, like in a conditional.

paramat commented on src/server.cpp:
```diff
@@ -2863,28 +2863,28 @@ std::wstring Server::handleChat(const std::string &name, const std::wstring &wna
+		}
+		case RPLAYER_CHATRESULT_KICK:
+			DenyAccess_Legacy(player->getPeerId(),
+					L"You have been kicked due to message flooding.");
```
1 indent is enough

paramat replied:
Sorry, as long as it's readable any amount of indentation is ok.

paramat commented on src/server.cpp:
```diff
@@ -2912,10 +2912,8 @@ std::wstring Server::handleChat(const std::string &name, const std::wstring &wna
-		line += L"> ";
-		line += wmessage;
+		line += narrow_to_wide(m_script->formatChatMessage(name,
+				wide_to_narrow(wmessage)));
```
1 indent is enough

paramat replied:
Sorry, as long as it's readable any amount of indentation is ok.

ClobberXD commented:
Rebased. Bumpity-bump :)

benrob0329 commented:
Does your code check for preceding backslashes for escaping? If not, it should. If so, then you should sanitize the chat before sending it through the `minetest.conf` setting.

ClobberXD commented on src/gettime.h:
```diff
@@ -19,7 +19,6 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 #pragma once
 
-#include "irrlichttypes.h"
```
I made this change when the chat message was formatted in the C++-side.

I haven't removed this, as this header needn't be included in this file and would increase compilation times for no reason.

ClobberXD commented:
> Does your code check for preceding backslashes for escaping?

No(t yet). This is exactly what DS-Minetest suggested, but is something don't think this is worth implementing, as this could make the Lua code unnecessarily long. Note however, that the actual chat-message can contain placeholders and won't be affected, because only the first instance of the placeholder will be replaced. Also, the message string itself isn't parsed, and is just plugged into the format string, in place of the first `@message` placeholder.

ClobberXD commented:
Rebased.
